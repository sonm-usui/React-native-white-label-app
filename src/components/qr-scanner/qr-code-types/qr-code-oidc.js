// @flow

import { type Url } from 'url-parse'
import isUrl from 'validator/lib/isURL'
import { stringify } from 'query-string'

import type {
  QrCodeOIDC,
  JWTAuthenticationRequest,
  QR_SCAN_STATUS,
} from '../type-qr-scanner'

import { QR_CODE_TYPES, SCAN_STATUS } from '../type-qr-scanner'
import { acceptHeaders, flatFetch } from '../../../common/flat-fetch'
import { schemaValidator } from '../../../services/schema-validator'
import { toUtf8FromBase64 } from '../../../bridge/react-native-cxs/RNCxs'
import { addBase64Padding } from '../../../common/base64-padding'
import { flatTryCatch } from '../../../common/flat-try-catch'
import { deepLinkAddress } from '../../../external-imports'
import { flatJsonParse } from '../../../common/flat-json-parse'
import { isValidUrl } from './qr-url'
import type { GenericObject } from '../../../common/type-common'

export function isValidOIDCQrCode(parsedUrl: Url): QrCodeOIDC | false {
  const { protocol, query, hostname } = parsedUrl

  if (!validInvitationUrlScheme.includes(protocol)) {
    return false
  }

  if (!deepLinkAddress || deepLinkAddress !== hostname) {
    return false
  }

  // validate query string params that needs to be present in OIDC link
  for (const requiredParamName of requiredQueryParams) {
    if (!query[requiredParamName]) {
      return false
    }
  }

  // validation for values of required query params
  const responseType = query.response_type
  if (responseType !== 'id_token') {
    return false
  }

  const [decodeError, clientId] = flatTryCatch(decodeURIComponent)(
    query.client_id
  )
  if (decodeError || !clientId || !isUrl(clientId)) {
    return false
  }

  const [requestDecodeError, requestUri] = flatTryCatch(decodeURIComponent)(
    query.request_uri
  )
  if (requestDecodeError || !requestUri || !isUrl(requestUri)) {
    return false
  }

  return {
    type: QR_CODE_TYPES.OIDC,
    version: '0.1',
    clientId,
    requestUri,
    responseType,
  }
}

export async function fetchValidateJWT(
  oidcAuthenticationCode: QrCodeOIDC
): Promise<[null | JWTAuthenticationRequest, null | string]> {
  const [error, jwtResponse] = await flatFetch(
    oidcAuthenticationCode.requestUri
  )
  if (error !== null || jwtResponse === null) {
    return [null, SCAN_STATUS.AUTH_REQUEST_DOWNLOAD_FAILED]
  }

  const [encodedHeader, encodedBody, encodedSignature] = jwtResponse
    .split('.')
    .map(addBase64Padding)

  // check if header is valid as per MSDK expectations
  const [decodeError, header] = await decodeToJSON(encodedHeader)
  if (decodeError !== null || header === null) {
    return [null, SCAN_STATUS.AUTH_REQUEST_INVALID_HEADER_DECODE_ERROR]
  }
  if (!schemaValidator.validate(jwtHeaderSchema, header)) {
    return [null, SCAN_STATUS.AUTH_REQUEST_INVALID_HEADER_SCHEMA]
  }

  // check if body is valid
  const [bodyDecodeError, body] = await decodeToJSON(encodedBody)
  if (bodyDecodeError !== null || body === null) {
    return [null, SCAN_STATUS.AUTH_REQUEST_INVALID_BODY_DECODE_ERROR]
  }
  if (!schemaValidator.validate(jwtBodySchema, body)) {
    const [fail] = await sendValidationErrorToClient(
      oidcAuthenticationCode.clientId,
      JSON.stringify(schemaValidator.errors),
      body
    )
    if (fail) {
      return [null, SCAN_STATUS.AUTH_REQUEST_INVALID_BODY_SCHEMA_AND_SEND_FAIL]
    }
    return [null, SCAN_STATUS.AUTH_REQUEST_INVALID_BODY_SCHEMA]
  }

  // TODO:KS Add functionality to add signature verification using libVcx
  // For now, signature may or may not be available, and we are going ahead
  // with requests even without signature. Once this changes, and we decide
  // to enforce signatures and signature validation, then we need to add logic
  // here to enforce signature verification using libVcx
  // check if signature is present
  // if (encodedSignature.length < 2) {
  //   return [null, SCAN_STATUS.AUTH_REQUEST_INVALID_SIGNATURE]
  // }

  return [
    {
      header,
      body,
      encodedSignature,
    },
    null,
  ]
}

async function decodeToJSON(
  encodedValue: string
): Promise<[null | typeof Error, null | Object]> {
  try {
    const decodedUtf = await toUtf8FromBase64(encodedValue, 'URL_SAFE')
    const decodedJSON = JSON.parse(decodedUtf)

    return [null, decodedJSON]
  } catch (e) {
    return [e, null]
  }
}

async function sendValidationErrorToClient(
  url: string,
  errors: string,
  body: Object
) {
  const state =
    !body.state || typeof body.state !== 'string'
      ? 'property state missing in payload'
      : body.state

  const response = {
    error: 'invalid_request',
    error_description: errors,
    state,
  }
  const [sendError, status] = await flatFetch(url, stringify(response), {
    'Content-Type': 'application/x-www-form-urlencoded',
  })

  return [!!sendError, !!status]
}

// only trust if scheme is https, http is not allowed
const validInvitationUrlScheme = ['https:']

// required query string params
const requiredQueryParams = ['response_type', 'client_id', 'request_uri']

const jwtHeaderSchema = {
  type: 'object',
  properties: {
    alg: { type: 'string', enum: ['Ed25519', 'none'] },
    typ: { type: 'string', enum: ['JWT'] },
    kid: { type: 'string', minLength: 2, maxLength: 1024 },
  },
  required: ['alg', 'typ'],
}

const jwtBodySchema = {
  type: 'object',
  properties: {
    iss: { type: 'string', minLength: 3 },
    response_type: { type: 'string', minLength: 2 }, // need to add ENUM for possible values of response_type
    client_id: { type: 'string', minLength: 5, maxLength: 4096 },
    scope: { type: 'string', minLength: 2 },
    state: { type: 'string', minLength: 2 },
    nonce: { type: 'string', minLength: 8 },
    response_mode: { type: 'string', enum: ['form_post', 'query'] },
    registration: {
      type: 'object',
      properties: {
        id_token_signed_response_alg: {
          type: 'array',
          minItems: 1,
          enum: [['Ed25519']],
        },
        request_object_signing_alg: { type: 'string' },
        jwks_uri: { type: 'string', minLength: 4, maxLength: 4096 },
      },
      required: ['id_token_signed_response_alg'],
    },
  },
  required: [
    'iss',
    'response_type',
    'client_id',
    'scope',
    'nonce',
    'response_mode',
    'registration',
    'state',
  ],
}

const query_param = 'request_uri='
const domain = 'openid://'

export function isValidOpenIDLink(data: string): boolean {
  return data.startsWith(domain)
}

export async function getOpenidLinkData(
  link: string
): Promise<[null | QR_SCAN_STATUS, GenericObject | null]> {
  if (!link.includes(query_param)) {
    return [SCAN_STATUS.INVALID_OPENID_QR_LINK, null]
  }

  const requestUri = link.split(query_param)[1]
  const parsedRequestUri = isValidUrl(requestUri)
  if (!parsedRequestUri) {
    return [SCAN_STATUS.INVALID_OPENID_QR_LINK, null]
  }

  const [, downloadedData] = await flatFetch(
    requestUri,
    undefined,
    acceptHeaders
  )
  if (!downloadedData) {
    return [SCAN_STATUS.INVALID_OPENID_QR_LINK, null]
  }

  const [, parsedData] = flatJsonParse(downloadedData)
  if (parsedData) {
    // if we get some json data, then return it
    return [null, parsedData]
  }

  return [SCAN_STATUS.INVALID_OPENID_QR_LINK, null]
}
