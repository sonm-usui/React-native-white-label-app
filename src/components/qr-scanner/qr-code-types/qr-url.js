// @flow
import urlParse, { type Url } from 'url-parse'

import type {
  AriesConnectionInvite,
  AriesOutOfBandInvite,
} from '../../../invitation/type-invitation'
import type {
  QrCodeOIDC,
  QR_SCAN_STATUS,
  QrCodeNonJsonUrl,
} from '../type-qr-scanner'

import { SCAN_STATUS, QR_CODE_TYPES } from '../type-qr-scanner'
import type { GenericObject } from '../../../common/type-common'
import { acceptHeaders, flatFetch } from '../../../common/flat-fetch'
import { flatJsonParse } from '../../../common/flat-json-parse'
import { isValidOIDCQrCode } from './qr-code-oidc'
import { getRequestRedirectionUrl } from '../../../bridge/react-native-cxs/RNCxs'
import { flattenAsync } from '../../../common/flatten-async'
import { isEncodedAriesConnectionInvitation } from '../../../invitation/kinds/aries-connection-invitation'
import { isEncodedAriesOutOfBandInvitation } from '../../../invitation/kinds/aries-out-of-band-invitation'
import { getBase64DecodedInvitation } from '../../../invitation/invitation-helpers'
import { schemaValidator } from '../../../services/schema-validator'
import { ephemeralProofRequestSchema } from '../../../proof-request/proof-request-qr-code-reader'
import { ephemeralCredentialOfferSchema } from '../../../claim-offer/ephemeral-claim-offer'

export function isValidUrl(urlQrCode: ?string): Url | false {
  if (!urlQrCode) {
    return false
  }

  const parsedUrl = urlParse(urlQrCode, {}, true)

  if (!validUrlScheme.includes(parsedUrl.protocol)) {
    return false
  }

  return parsedUrl
}

export async function getUrlData(
  parsedUrl: Url,
  url: string
): Promise<
  [
    null | QR_SCAN_STATUS,
    (
      | null
      | GenericObject
      | AriesConnectionInvite
      | AriesOutOfBandInvite
      | QrCodeOIDC
      | QrCodeNonJsonUrl
    )
  ]
> {
  // if we get url qr code, then there are three ways as of now that ConnectMe supports
  // to get data from url qr code

  // Three ways are to get data directly from URL query params

  // 1. get aries invitation data using url qr code
  const ariesConnectionInvite = await isEncodedAriesConnectionInvitation(
    parsedUrl
  )
  if (ariesConnectionInvite) {
    return [null, ariesConnectionInvite]
  }

  const ariesOutofbandConnectionInvite = await isEncodedAriesOutOfBandInvitation(
    parsedUrl
  )
  if (ariesOutofbandConnectionInvite) {
    return [null, ariesOutofbandConnectionInvite]
  }

  // 2. get OIDC authentication data from url
  const oidcQrCode = isValidOIDCQrCode(parsedUrl)
  if (oidcQrCode) {
    return [null, oidcQrCode]
  }

  // 3. get out of band invitation data from URL query parameter
  const outOfBandInvite = await isEncodedAriesOutOfBandInvitation(parsedUrl)
  if (outOfBandInvite) {
    return [null, outOfBandInvite]
  }

  // didcomm url cannot be requested -> return
  if (url.startsWith('didcomm:')) {
    return [SCAN_STATUS.INVALID_URL_QR_CODE, null]
  }

  // if there is no data available in url params, then try to download data
  // from the passed url and check if we get data from url

  // 4. get URL data. do not follow redirection. just return new url
  let [, redirectionUrl] = await flattenAsync(getRequestRedirectionUrl)(url)
  const parsedRedirectionUrl = isValidUrl(redirectionUrl || '')
  if (redirectionUrl && parsedRedirectionUrl) {
    // BCGov links contains message

    if (parsedRedirectionUrl.query) {
      const message =
        parsedRedirectionUrl.query['m'] || parsedRedirectionUrl.query['d_m']
      if (message) {
        const data = await getBase64DecodedInvitation(message)

        // is ephemeral proof request
        if (schemaValidator.validate(ephemeralProofRequestSchema, data)) {
          return [
            null,
            { type: QR_CODE_TYPES.EPHEMERAL_PROOF_REQUEST_V1, data: data },
          ]
        }

        // is ephemeral credential offer
        if (schemaValidator.validate(ephemeralCredentialOfferSchema, data)) {
          // Ephemeral Credential Offer can be converted into Out-of-Band invitation without handshake
          // as they have similar meaning / handling flow
          // so we can easily process it using already implemented functionality
          return [
            null,
            { type: QR_CODE_TYPES.EPHEMERAL_CREDENTIAL_OFFER, data: data },
          ]
        }
      }
    }

    // else handle link again
    // downloaded data contains url - get its data
    return await getUrlData(parsedRedirectionUrl, redirectionUrl)
  }

  // 5. download data and get a valid json object
  const [downloadErr, downloadedData] = await flatFetch(
    url,
    undefined,
    acceptHeaders
  )
  if (downloadedData) {
    // we are able to get data from url
    // now we need to verify that data is a valid json
    const [, parsedData] = flatJsonParse(downloadedData)

    if (parsedData) {
      // if we get some json data, then return it
      return [null, parsedData]
    }

    const urlInvitationData = isValidUrl(downloadedData)
    if (urlInvitationData) {
      // downloaded data contains url - get its data
      return await getUrlData(urlInvitationData, downloadedData)
    }

    // since we know that we got some downloadedData
    // but this downloaded data is not json
    // so we return with type non-json

    return [
      null,
      { type: QR_CODE_TYPES.URL_NON_JSON_RESPONSE, data: downloadedData },
    ]
  }
  if (downloadErr) {
    // $FlowFixMe
    return [downloadErr.message, null]
  }

  // if we reach to this point, that means that we could not get data from url
  // either by downloading data behind url, or by extracting data from url itself
  // so the url qr code is invalid
  return [SCAN_STATUS.INVALID_URL_QR_CODE, null]
}

export const validUrlScheme = ['https:', 'http:', 'id.streetcred:', 'didcomm:']
