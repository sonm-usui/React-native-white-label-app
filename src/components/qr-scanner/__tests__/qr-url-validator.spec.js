// @flow

import { isValidUrl } from '../qr-code-types/qr-url'

describe('fn:isValidUrlQrCode', () => {
  it('should return true, if protocol is http', () => {
    const urlQrCode = `http://xya.com/dev`
    expect(isValidUrl(urlQrCode)).toMatchSnapshot()
  })

  it('should return true, if protocol is https', () => {
    const urlQrCode = `https://xya.com/dev`
    expect(isValidUrl(urlQrCode)).toMatchSnapshot()
  })

  it('should return false, if protocol is not valid', () => {
    const urlQrCode = `ftp://xya.com/dev`
    expect(isValidUrl(urlQrCode)).toBe(false)
  })

  it('should return false, if protocol is not valid', () => {
    const urlQrCode = `file:///xya.com/dev`
    expect(isValidUrl(urlQrCode)).toBe(false)
  })
})
