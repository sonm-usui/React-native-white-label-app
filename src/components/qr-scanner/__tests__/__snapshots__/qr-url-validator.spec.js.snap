// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fn:isValidUrlQrCode should return true, if protocol is http 1`] = `
Object {
  "auth": "",
  "hash": "",
  "host": "xya.com",
  "hostname": "xya.com",
  "href": "http://xya.com/dev",
  "origin": "http://xya.com",
  "password": "",
  "pathname": "/dev",
  "port": "",
  "protocol": "http:",
  "query": Object {},
  "slashes": true,
  "username": "",
}
`;

exports[`fn:isValidUrlQrCode should return true, if protocol is https 1`] = `
Object {
  "auth": "",
  "hash": "",
  "host": "xya.com",
  "hostname": "xya.com",
  "href": "https://xya.com/dev",
  "origin": "https://xya.com",
  "password": "",
  "pathname": "/dev",
  "port": "",
  "protocol": "https:",
  "query": Object {},
  "slashes": true,
  "username": "",
}
`;
